### ISSUE TITLE
How can a mix of polygon and bounding box annotations be used to train an object detection model in YOLOv8?

### ISSUE THREAD

**Issue by Camerabiologist** [ultralytics/ultralytics#1502]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have tried to find an answer to this question, but didn't find a clear answer. For training a custom object detection model, I have annotated a dataset of images with my designated object. The annotation was mostly done using the rectangular bounding boxes, but about 20% of my images, I opted to use a polygon, as this would reduce a significant amount of background noise in the training dataset. Now I'm trying to implement my dataset with Yolov8, but my annotating tool (cvat.ai) doesn't correctly export my polygonal labels to .txt label files, as opposed to the bounding boxes, which are exporting just fine. This lead me to the question whether it's even possible in the current Yolov8 to train a model using polygon annotation, or a mix of annotations, like 80% bounfing boxes, 20% polygons. If anyone could help me understand this problem, it would be greatly appreciated.

Also, part of the problem is the failed exports from cvat.ai, so if you have experience with that annotation software, I would be happy to know how to effectively export polygon annotation. Of course I will also try to solve the problem with the cvat people.

If any additional information would be useful to answer my question, please do let me know.

Thank you!

**Comment by JustasBart** [comment]
Hi @Camerabiologist sounds to me like you might be able to create just a simple Python based tool that could read in polygon based labels and use something like the `x,y,w,h = cv2.boundingRect(contour)` function to extract your desired rectangles:

![image](https://user-images.githubusercontent.com/40023722/226284265-b7616e43-178a-4790-9633-f9f2e697868f.png)

Then simply just save them by keeping the class id untouched and writing out the rectangular label file.
Keep in mind though that the labels are meant to be in a range of 0.0 to 1.0 on the both X/Y coordinates so you might also have to do a few conversions.

Perhaps ask ChatGPT to walk you through this?

Good luck! :rocket:

**Comment by Camerabiologist** [comment]
Thank you for your response! The goal of my polygons, though, was to reduce the amount of background in the labels. Specifically, I am looking at identifying a particular plant species, which can often co-occur in the same area as other plants. Therefore I sometimes deemed it necessary to turn away from the rectangles. Therefore, if there is any way in which i can avoid changing the labels back to rectangles, I would prefer that. Is that possible?

**Comment by JustasBart** [comment]
@Camerabiologist Oh, I see what you mean, but no, unfortunately there isn't. And that's because the Object Detection model is very different than that of the Segmentation model, the two are very much so incompatible with each other and they can therefore not be used directly like you're describing. One of the issues with this for instance is the loss function, there's a loss function for the box (Object Detection) and another loss function for the Mask (Segmentation) so even if you could somehow train on both datasets at the same time you would run into the issue that now your data would become bias as there's only ~20% of the data that has the mask information versus just the bounding box.

Sorry to break out the bad news here, good luck! :rocket:

**Comment by Camerabiologist** [comment]
Alright, thanks for clearing that up for me! 

I am curious now though, could I make the current dataset work with a Segmentation model? Or would I then need to convert the whole dataset to polygon annotations?

**Comment by Camerabiologist** [comment]
Thanks again for your help, I will be researching the best continuation of my project then!

**Comment by NussairHroub** [comment]
@Camerabiologist 
Hi, I would like to know if you find a solution to your problem. I am trying to convert a text file that contains a bounding box format to a text file of a polygon format but still do not know how the polygon is represented.
Thanks

**Comment by glenn-jocher** [comment]
@NussairHroub hello! Regarding your question on converting bounding box format to polygon format, you could modify the conversion code to output the polygon vertices in the text file. It is important to note that a polygon is represented as a series of vertices, with each vertex consisting of an x and y coordinate. This informs the shape of the polygon. The polygon vertices can be generated from the bounding box coordinates by selecting four points on the perimeter of the rectangular bounding box and using them as the polygon vertices. The vertices may be generated in any order, as long as they are consistently ordered for all polygons in the dataset. I hope this helps, let me know if there's anything else I can assist you with!

**Comment by NussairHroub** [comment]
> glenn-jocher

Thanks for your response. I have converted the bounding boxes to polygons. I have one more question, please. In the dataset I have, some of the images are not annotated since they do not contain the target object for segmentation. Is it better to keep those instances, or remove them or it is indifferent?
Thanks again,

**Comment by SBhat2615** [comment]
@Camerabiologist 
Hi, I would like to know about how did you go with your project. How to proceed in finding ROI where the objects are curvilinear ?? Even though i have 1 class, because of the shape it might not be best way to proceed with the bounding box is what i observed. 
I have a segmented image of the object. Can i do something with that ??
If i draw bounding boxes in some images, there are many empty spaces. Structures are also hard to differentiate.
It would be very helpful if anyone can help.
Thanks.

**Comment by glenn-jocher** [comment]
@SBhat2615 Thanks for reaching out. Dealing with curvilinear objects can indeed be a challenge for conventional bounding box approaches, as they may lead to large areas of irrelevant background information being included.

One way to handle this situation could be through instance segmentation, which associates pixels with a specific instance of an object rather than a bounding box, thus accommodating more complex shapes efficiently. However, note that the Ultralytics YOLOv8 implementation is primarily for object detection (i.e., standard rectangle-based bounding boxes).

Your segmented images could certainly be useful, perhaps in a segmentation-based model rather than a detection-based one. However, you'd need to ensure these segmentations are accurate and reliable because these would be the ground truth annotations your model is trained on.

It sounds like your problem could be more suited to segmentation than detection, given the nature of your objects and the issues you're facing with bounding boxes. That being said, each approach comes with its own challenges and trade-offs.

For curvilinear objects, it may be worth exploring algorithms specifically designed for such tasks, beyond the traditional object detection or segmentation approaches.

I hope this guidance is useful, and I wish you luck in your project!

**Comment by SBhat2615** [comment]
> @SBhat2615 Thanks for reaching out. Dealing with curvilinear objects can indeed be a challenge for conventional bounding box approaches, as they may lead to large areas of irrelevant background information being included.
> 
> One way to handle this situation could be through instance segmentation, which associates pixels with a specific instance of an object rather than a bounding box, thus accommodating more complex shapes efficiently. However, note that the Ultralytics YOLOv8 implementation is primarily for object detection (i.e., standard rectangle-based bounding boxes).
> 
> Your segmented images could certainly be useful, perhaps in a segmentation-based model rather than a detection-based one. However, you'd need to ensure these segmentations are accurate and reliable because these would be the ground truth annotations your model is trained on.
> 
> It sounds like your problem could be more suited to segmentation than detection, given the nature of your objects and the issues you're facing with bounding boxes. That being said, each approach comes with its own challenges and trade-offs.
> 
> For curvilinear objects, it may be worth exploring algorithms specifically designed for such tasks, beyond the traditional object detection or segmentation approaches.
> 
> I hope this guidance is useful, and I wish you luck in your project!

Thanks a lot for your response.
Yes i've looked into to the article regarding instance, polygon and semantic segmentation.
I've tried a Attention U-Net segmentation model, which gives a decent result. Now i want find out ROI, which can be basically a bounding box. Is there a better to proceed ??
Also, for the purpose of training i have a hand annotated ground truth segmented images other than the U-Net model.

**Comment by SBhat2615** [comment]
> Hello @SBhat2615,
> 
> Glad to hear you're making progress with Attention U-Net for segmentation!
> 
> When it comes to finding the ROI, a common method after performing a segmentation task is to compute bounding boxes around individual connected components or clusters in the segmented image. This is a standard approach and should work well if the segments in your output image are well isolated.
> 
> In terms of training, it's always better to use a variety of ground truth data if available. Using both hand-annotated segmented images and U-Net generated segmentations could provide a more robust training, as long as the quality of both sets of annotations is good.
> 
> Remember to ensure that the precision and recall of your model are evaluated using a fair metric, as this can greatly affect the performance of your object detection or segmentation task.
> 
> Best of luck with your project!

Cool !! I wrote a script to compute the bounding boxes around individual components using the segmented images. Will proceed in the same way then. Thanks a lot for your guidance and quick response !!

**Comment by glenn-jocher** [comment]
@SBhat2615,

Great to hear that you've already made progress with computing bounding boxes around the individual components of your segmented images! It sounds like you're on the right track. 

Your approach to use a mix of hand-annotated and U-Net generated segmentations for training should make your model robust. Just remember to continually assess the quality of your annotations, as this directly impacts the training effectiveness.

Finally, continuously monitor your model's precision and recall to ensure optimal performance. These metrics are key to understanding how well your model is doing.

Continue the great work and don't hesitate to ask if you have any more questions or need further guidance. Good luck with your project!

**Comment by DavidFlowbender** [comment]
This is my first time annotating and I used a polygon shape annotation instead of bounding boxes. I don't want to go through the stress of drawing bounding boxes again. Is there a way to covert the polygon shape to bounding boxes, since I want to train my model with YOLO

**Comment by tPiffrader** [comment]
> This is my first time annotating and I used a polygon shape annotation instead of bounding boxes. I don't want to go through the stress of drawing bounding boxes again. Is there a way to covert the polygon shape to bounding boxes, since I want to train my model with YOLO

Maybe this can help: https://docs.opencv.org/4.x/da/d0c/tutorial_bounding_rects_circles.html

**Comment by DavidFlowbender** [comment]
Thank you

**Comment by glenn-jocher** [comment]
@DavidFlowbender hello Tobias,

Thank you for reaching out with your question about converting polygon shape annotations to bounding boxes for training with YOLOv8.

The good news is that you don't have to undergo the process of redrawing bounding boxes if you've already annotated your dataset with polygons. YOLOv8 requires annotations to be in a specific format that consists of the object class index and the bounding box coordinates.

To convert your polygon annotations to this format, the typical approach is to calculate the axis-aligned minimum bounding box for each polygon. This process involves determining the smallest rectangle that can fully contain the polygon while its sides are aligned with the x and y axes. For each polygon, you would find the minimum and maximum x and y coordinates among the vertices of the polygon. These coordinates will correspond to the top-left and bottom-right corners of the bounding box.

After calculating these bounding boxes, you'll need to normalize the coordinates according to the dimensions of the input images for YOLOv8, and ensure the class labels are correctly assigned. This process is usually handled by an annotation conversion script or tool which can transform the polygon annotation data into the YOLOv8 format.

Keep in mind that bounding boxes are simpler geometric shapes compared to polygons and may not represent the exact shape of the objects with high fidelity, especially if the objects are of a non-rectangular nature. However, if the polygons are relatively compact and the objects are well-represented by rectangles, the transformation to bounding boxes should be appropriate for YOLOv8 training.

We understand the importance of having a streamlined and efficient annotation process. We hope this information helps you convert your polygons to bounding boxes without additional stress. If you need any more assistance, please feel free to reach out.

Best regards and happy training with YOLOv8!

**Comment by pratikshac15** [comment]
@glenn-jocher Hi Glenn, I was wondering if I used both rectangular and polygon bounding boxes in one dataset would it affect the accuracy of the model? and how the model detect one class of objects trained with polygon bounding boxes and other classes with the rectangular box.
Thanks.

**Comment by glenn-jocher** [comment]
@pratikshac15 hi there! 😊

Mixing rectangular and polygon bounding boxes in one dataset for training with YOLOv8 won't directly affect the model's accuracy. YOLOv8 uses rectangular bounding boxes for detection, so polygons would need to be converted to their bounding rectangle equivalents before training.

For classes annotated with polygons, converting them to bounding rectangles might include some background noise, depending on the object's shape. This could potentially affect how well the model learns those particular classes compared to those annotated with precise rectangular boxes. However, the impact largely depends on the dataset's specifics and the objects' shapes.

In practice, ensure all annotations are converted to the format YOLO expects (rectangular bounding boxes), and the model should learn to detect objects regardless of the original annotation shape. Just keep an eye on the performance metrics during training to catch any significant discrepancies early on!

Hope this helps! Let me know if you have more questions.

**Comment by pratikshac15** [comment]
@glenn-jocher Thank you Glenn, This was helpful. One of my class is thin fibers and has irregular shape. This is always taking a lot of background during rectangular bounding box labeling. and because of that the accuracy of fiber class is always less than 80% other classes are 95-98% range.

Any suggestions ?

**Comment by glenn-jocher** [comment]
@pratikshac15 Hi there! 😊 I'm glad you found the information helpful. For thin and irregularly shaped objects like fibers, capturing too much background in rectangular bounding boxes can indeed be challenging and might affect your model's accuracy for that class.

One approach to consider is augmenting your dataset specifically for the fiber class to include more varied backgrounds and orientations. This can help the model become more robust to the background noise. Additionally, playing with the anchor box sizes in YOLOv8's configuration might help the model better adapt to the aspect ratio of thin fibers.

Here's a quick example of how you might adjust the anchor sizes in your YOLOv8 config file:

```yaml
anchors:
  - [10,13, 16,30, 33,23]  # Small anchors
  - [30,61, 62,45, 59,119]  # Medium anchors
  - [116,90, 156,198, 373,326]  # Large anchors
```

You might want to experiment with smaller anchors for the fiber class to see if it improves detection.

Keep experimenting, and best of luck with your model!

**Comment by glenn-jocher** [comment]
Thanks for sharing the Roboflow article. Ultralytics YOLO models use rectangular bounding boxes for object detection as shown in our Data Collection and Annotation guide, while polygons are supported for instance segmentation tasks. For detection tasks, polygon annotations need to be converted to their minimal enclosing rectangles for YOLO compatibility. Let us know if you need help with annotation format conversion!

**Comment by MarcioPalmutti** [comment]
Dear Jocher,
Without wanting to be controversial, I apologize in advance for my insistence and ignorance!
However, in the article I cited above, at no point does it mention that we need to convert the polygons into rectangles.
Note that you yourself commented on 11/15/2023 on the possibility of using the polygon instead of the rectangle: https://github.com/ultralytics/ultralytics/issues/1502#issuecomment-2727256167
Therefore, I rephrase my question: can I use polygon annotations in YOLO detection?
If not, what is the best way to use YOLO to locate small objects (seeds) in an image, detection or segmentation?

**Comment by MarcioPalmutti** [comment]
Dear Jocher,
Without wanting to be controversial, I apologize in advance for my insistence and ignorance!
However, in the article I cited above, at no point does it mention that we need to convert the polygons into rectangles.
Note that you yourself commented on 11/15/2023 on the possibility of using the polygon instead of the rectangle: https://github.com/ultralytics/yolov5/issues/10452#issuecomment-1811981887
Therefore, I rephrase my question: can I use polygon annotations in YOLO detection?
If not, what is the best way to use YOLO to locate small objects (seeds) in an image, detection or segmentation?

**Comment by glenn-jocher** [comment]
Hi there! 😊 YOLO detection models (like YOLOv8) use **rectangular bounding boxes** for object detection, while segmentation models use polygons/masks. For detection tasks, polygon annotations must be converted to their minimal enclosing rectangles as shown in our [Data Annotation Guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/#types-of-data-annotation). 

For small objects like seeds, detection works well if objects are discrete. If precise shape delineation is critical (e.g., overlapping objects), consider using YOLO **segmentation models** with polygon annotations as described in our [Segmentation Tasks Documentation](https://docs.ultralytics.com/tasks/segment/). Let us know if you need further guidance!

**Comment by MarcioPalmutti** [comment]
Sorry,
But what do you think about your comment on 11/15/2023, where you state that in YOLOv5 it is possible to use polygon annotation to train YOLO, where it can offer more accurate detection than with rectangles?
You said:
"YOLOv5 supports polygon segmentation annotations for object detection. You can use polygon annotation to train the model, and it can offer more accurate detection than rectangle annotation in certain scenarios. You can refer to the Ultralytics documentation for detailed guidance on using polygon annotation for training. Good luck with your training!"
As I said, I don't want to be controversial at all, but if polygon annotation is possible for detection, it would be very important for my use case, especially for object contouring and background exclusion.
Thanks!

**Comment by glenn-jocher** [comment]
@MarcioPalmutti hi there! 😊 YOLO detection models (like YOLOv5/YOLOv8) use **rectangular bounding boxes** for object detection, while segmentation models use polygons/masks. The comment you referenced was likely referring to segmentation capabilities. For detection tasks, polygon annotations must be converted to their minimal enclosing rectangles as shown in our [Data Annotation Guide](https://docs.ultralytics.com/guides/data-collection-and-annotation/#types-of-data-annotation). 

If precise object contouring is critical, we recommend using YOLO **segmentation models** with polygon annotations as described in our [Segmentation Tasks Documentation](https://docs.ultralytics.com/tasks/segment/). Let us know if you need further clarification! 🚀
